### PR TITLE
drawRect_toCanvas() added

### PR DESCRIPTION
**Pull request by Akshay2131**
Overview

 Fixed the issue "Bitmap-backed canvas.drawRect fails with paint" #7188 

 Proposed Changes

 Earlier, the test was failing in Robolectric with the issue : 

`assertThat(bitmap.getPixel(0, 0)).isEqualTo(Color.WHITE);
Expected :-1
Actual   :-16777216`

It was failing with:

`canvas.drawRect(0, 0, 10, 10, paint);`

It was failing to draw a rectangle on the canvas.
 So, bitmap.getPixel(0, 0) was giving the old pixel value.

I changed the statement from: 

`canvas.drawRect(0, 0, 10, 10, paint);` 

to this:

`canvas.drawRect(new Rect(0, 0, 10, 10), paint);`

Now, it is drawing a rectangle on the canvas, 
Now, the test is working properly.
  
 I have added the test to ShadowBitmapTest.java file

### CODE REVIEW

File: robolectric/src/test/java/org/robolectric/shadows/ShadowBitmapTest.java [hunk 1]
```diff
@@ -822,4 +822,19 @@ private void createScaledBitmap_drawOnScaled(boolean filter) {
     p.setColor(Color.BLACK);
     canvas.drawRect(new Rect(0, 0, 32, 32), p);
   }
+
+  @Test
+  public void drawRect_toCanvas() {
```
Comment (utzcoz) [inline robolectric/src/test/java/org/robolectric/shadows/ShadowBitmapTest.java:827]: 1. I think this test should moved to `ShadowCanvasTest.java`.
2. This test only ensure `Canva#drawRect(Rect rect, Paint paint)` works on Robolectric, but it doesn't ensure `Canvas#drawRect(float left, float top, float right, float bottom, Paint paint)` works as expected, and it is shown by origin issue.
3. From https://github.com/robolectric/robolectric/blob/183437a0b45ff97f2df3128ec01896ffaef5e737/shadows/framework/src/main/java/org/robolectric/shadows/ShadowCanvas.java#L243-L256, the shadow of `Canva#drawRect(float left, float top, float right, float bottom, Paint paint)` doesn't update target `Bitmap` content as `Canvas#drawRect(Rect rect, Paint paint)` did. IMO, you could improve this shadow implementation and add related test for it at `ShadowCanvasTest.java`.

**Review by utzcoz — CHANGES_REQUESTED** [review]

File: robolectric/src/test/java/org/robolectric/shadows/ShadowBitmapTest.java (hunk 1 above)
Comment (hoisie) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowBitmapTest.java:827]: Yeah, the goal here was to fix Robolectric in a way that the original test passes.

There were some nuances to this issue that I did not think about initially.

There are two variants of Canvas.drawRect-- one if them is `drawRect(Rect r, Paint paint)`, and the other is `drawRect(float left, float top, float right, float bottom, Paint paint)`. Currently, in Robolectric,  [drawRect(Rect r, Paint paint)](https://github.com/robolectric/robolectric/blob/master/shadows/framework/src/main/java/org/robolectric/shadows/ShadowCanvas.java#L252-L255) has a more correct shadow implementation, in that it calls into ShadowBitmap.drawRect to do actual drawing.

The `drawRect(float left, float top, float right, float bottom, Paint paint)` does not call into ShadowBitmap, so it is effectively a no-op. The goal is to update it to call into ShadowBitmap to do some drawing.

One nuance is that `Rect` in Android is a rect of ints, and `drawRect(float left, float top, float right, float bottom, Paint paint)` takes floats. So I don't think calling the existing ShadowBitmap.drawRect will work here (unless all of the floats have 0 in the fractional part).

One possibility is to implement a new ShadowBitmap.drawRect that takes a RectF and uses AWT:

```
  void drawRect(RectF r, Paint paint) {
    if (bufferedImage == null) {
      return;
    }
    Graphics2D graphics2D = bufferedImage.createGraphics();
    Rectangle2D r2d = new Rectangle2D.Float(r.left, r.top, r.right - r.left, r.bottom - r.top);
    graphics2D.draw(r2d);
    graphics2D.dispose();
  }
``` 

Then you could call that from ShadowCanvas using:

```
    if (targetBitmap != null) {
      ShadowBitmap shadowTargetBitmap = Shadows.shadowOf(targetBitmap);
      shadowTargetBitmap.drawRect(new RectF(left, top, right, bottom), paint);
    }
```

I vaguely recall reason that `ShadowBitmap.drawRect(Rect r, Paint paint)` does not use Graphics2d is because Graphics2d has slight behavior differences. I remember something like if you drawRect with something like 0x00ffffff, which is an 100% traslucent white color, Graphics2d converts it to `0`, and this broke some tests. At least I think that's what it is. Maybe I should capture that in CTesque BitmapTest.

Comment (Akshay2131) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowBitmapTest.java:827]: For this to work, i  need to create a new method in ShadowBitmap.java with prototype as: 
`void drawRect(int left, int top, int right, int bottom, Paint paint);`
 and then call to this method from shadowCanvas.java

Can i go with it?
what u think @utzcoz

Comment (Akshay2131) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowBitmapTest.java]: I have added a new method named drawRect in ShadowBitmap.java as it was required.
Could you review the changes..

File: robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java [hunk 2]
```diff
@@ -529,4 +530,19 @@ public void restoreToCount() {
 
     assertThat(canvas.getSaveCount()).isEqualTo(save1);
   }
+
```
Comment (utzcoz) [inline robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java:533]: We also need a test for `Canvas#drawRect(RectF rect, Paint paint)`.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java [hunk 3]
```diff
@@ -899,4 +903,27 @@ void setBufferedImage(BufferedImage bufferedImage) {
   BufferedImage getBufferedImage() {
     return bufferedImage;
   }
+
+  void drawRect(RectF r, Paint paint) {
```
Comment (utzcoz) [inline shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: Could you help to move this method under previous `drawRect`?

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java [hunk 4]
```diff
@@ -899,4 +903,27 @@ void setBufferedImage(BufferedImage bufferedImage) {
...
+      return;
+    }
+    int[] pixels = ((DataBufferInt) bufferedImage.getRaster().getDataBuffer()).getData();
+
+    Graphics2D graphics2D = bufferedImage.createGraphics();
+    Rectangle2D r2d = new Rectangle2D.Float(r.left, r.top, r.right - r.left, r.bottom - r.top);
+    graphics2D.draw(r2d);
+    graphics2D.dispose();
+
+    double bottom=r2d.getY()+r2d.getHeight();
```
Comment (utzcoz) [inline shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:918]: Please follow https://github.com/robolectric/robolectric/wiki/Robolectric's-code-style to format code style locally.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java [hunk 5]
```diff
@@ -899,4 +903,27 @@ void setBufferedImage(BufferedImage bufferedImage) {
...
     return bufferedImage;
   }
+
+  void drawRect(RectF r, Paint paint) {
+    if (bufferedImage == null) {
+      return;
+    }
+    int[] pixels = ((DataBufferInt) bufferedImage.getRaster().getDataBuffer()).getData();
+
+    Graphics2D graphics2D = bufferedImage.createGraphics();
```
Comment (utzcoz) [inline shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:913]: Maybe we need set paint color to `graphics2D` before `draw`:  `graphics2D.setColor(paint.getColor());`.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java [hunk 6]
```diff
@@ -899,4 +903,27 @@ void setBufferedImage(BufferedImage bufferedImage) {
...
+    graphics2D.dispose();
+
+    double bottom=r2d.getY()+r2d.getHeight();
+    double right=r2d.getX()+r2d.getHeight();
+    if (r2d.getX() == 0 && r2d.getY() == 0 && right == r2d.getWidth()) {
+      Arrays.fill(pixels, 0, (int)(r2d.getWidth() * bottom), paint.getColor());
+      return;
+    }
+    for (double y = r2d.getY(); y < bottom; y++) {
+      Arrays.fill(
```
Comment (utzcoz) [inline shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:925]: I think above `Graphics2D` has changed pixels content and we don't need do it again.

**Review by utzcoz — CHANGES_REQUESTED** [review]

File: robolectric/src/test/java/org/robolectric/shadows/ShadowBitmapTest.java (hunk 1 above)
Comment (utzcoz) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowBitmapTest.java]: @Akshay2131 I have posted new reviewing content.

File: robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java (hunk 2 above)
Comment (utzcoz) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java:533]: It's also recommended to add test to cover corner case pointed out by @hoisie at https://github.com/robolectric/robolectric/pull/7197#discussion_r839174063.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java (hunk 3 above)
Comment (Akshay2131) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: yeah, i will do it..

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java (hunk 4 above)
Comment (Akshay2131) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:918]: i had forgot to implement it..
done..

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java (hunk 5 above)
Comment (Akshay2131) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:913]: Actually, earlier i had tried it with the same statement `graphics2D.setColor(paint.getColor());`.
But, it was giving some error..
Now, I look into it once again..

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java (hunk 3 above)
Comment (hoisie) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: I'd recommend just having this method be:

```
  void drawRect(RectF r, Paint paint) {
    if (bufferedImage == null) {
      return;
    }
    Graphics2D graphics2D = bufferedImage.createGraphics();
    Rectangle2D r2d = new Rectangle2D.Float(r.left, r.top, r.right - r.left, r.bottom - r.top);
    graphics2D.draw(r2d);
    graphics2D.dispose();
  }
```

And not mucking with the underlying buffers.

File: robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java (hunk 2 above)
Comment (hoisie) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java:533]: I would probably recommend not worrying about this corner case (e.g. painting with invisible colors, e.g. 0x00ffffff). It's an edge case that is okay if it's not supported. We can look into it later if necessary.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java (hunk 5 above)
Comment (Akshay2131) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:913]: graphics2D.setColor(Color color) accepts argument of Color type, whereas paint.getColor() returns int value.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java (hunk 3 above)
Comment (Akshay2131) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: @hoisie the above code is not working as it is not using any color to draw it and it is failing tests..
So, we have to add either one statement into it as like : `graphics2D.setColor(paint.getColor());`
or we can fill pixels as i have done it, but it is not efficient..
I think we should add the above statement to set color for drawing..
I tested it and it's working..
There is one issue that graphics2D.setColor(Color color) accepts argument of Color type, whereas paint.getColor() returns int value.
I am looking to add it and then it willl work fine..

Comment (utzcoz) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: @Akshay2131 what about trying `new Color(paint.getColor())`? The `Color` has a constructor with int color value as input: https://docs.oracle.com/javase/8/docs/api/java/awt/Color.html#Color-int-.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java (hunk 5 above)
Comment (utzcoz) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:913]: See https://github.com/robolectric/robolectric/pull/7197#discussion_r840598262.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java (hunk 3 above)
Comment (Akshay2131) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: I am thinking to do the same either using constructor or method..
I will add a constructor which takes int color value as input..

Comment (Akshay2131) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: Yeah, it's working properly..

File: robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java (hunk 2 above)
Comment (Akshay2131) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java:533]: @utzcoz Is it for `Canvas#drawRect(RectF rect, Paint paint)` or `Canvas#drawRect(Rect rect, Paint paint)`..

Comment (Akshay2131) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java:533]: I think it should be for `Canvas#drawRect(Rect rect, Paint paint)`.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java (hunk 3 above)
Comment (hoisie) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: Whoops, this should work I believe:

  void drawRect(RectF r, Paint paint) {
    if (bufferedImage == null) {
      return;
    }
    Graphics2D graphics2D = bufferedImage.createGraphics();
    android.graphics.Color color = android.graphics.Color.valueOf(paint.getColor());
    graphics2D.setColor(new Color(color.red(), color.green(), color.blue(), color.alpha()));
    graphics2D.fill(new Rectangle2D.Float(r.left, r.top, r.right - r.left, r.bottom - r.top));
    graphics2D.dispose();
  }

Comment (Akshay2131) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: I added a line in the code you suggested earlier, and it's working now..
The code is as: 

`  void drawRect (RectF r, Paint paint) {

    if (bufferedImage == null) {
      return;
    }
    Graphics2D graphics2D = bufferedImage.createGraphics();
    Rectangle2D r2d = new Rectangle2D.Float(r.left, r.top, r.right - r.left, r.bottom - r.top);
    graphics2D.setColor(new Color(paint.getColor()));
    graphics2D.draw(r2d);
    graphics2D.dispose();
  } `

Comment (Akshay2131) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: I think the above code you suggested would also work..

Comment (Akshay2131) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: could you tell which additional test is to be added as pointed above by @utzcoz previously..

File: robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java (hunk 2 above)
Comment (utzcoz) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java:533]: @Akshay2131 I think it is for `Canvas#drawRect(RectF rect, Paint paint)`(https://developer.android.com/reference/android/graphics/Canvas#drawRect(android.graphics.RectF,%20android.graphics.Paint)). If you found Robolectric missed tests for other variants of `Canvas#drawRect`, it's recommended to add tests for them.

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java (hunk 3 above)
Comment (utzcoz) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: > I vaguely recall reason that ShadowBitmap.drawRect(Rect r, Paint paint) does not use Graphics2d is because Graphics2d has slight behavior differences. I remember something like if you drawRect with something like 0x00ffffff, which is an 100% translucent white color, Graphics2d converts it to 0, whereas in real Android this remains '0x00ffffff', and this broke some tests. At least I think that's what it is. Maybe I should capture that in CTesque BitmapTest.

Is this corner case pointed out by @hoisie. @hoisie can accept we don't add test for at this PR and check it at later PR. It depends on you to add it or not. All occasions are good to me.

**Comment by utzcoz** [comment]
@Akshay2131 It's recommend to squash all your commits locally and push it to this PR for further review and CI checking. :)

File: shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java (hunk 3 above)
Comment (hoisie) [reply shadows/framework/src/main/java/org/robolectric/shadows/ShadowBitmap.java:907]: Yeah it's fine not to handle or have a test for this corner case. We should try to defer as much to AWT as possible for this implementation.

File: robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java [hunk 7]
```diff
@@ -529,4 +530,36 @@ public void restoreToCount() {
...
+    Paint paint = new Paint();
+    paint.setStyle(Paint.Style.FILL);
+    paint.setAntiAlias(true);
+    paint.setColor(Color.WHITE);
+    canvas.drawRect(0, 0, 10, 10, paint);
+    assertThat(bitmap.getPixel(0, 0)).isEqualTo(Color.WHITE);
+  }
+
+  @Test
+  public void draw_Rect() throws Exception {
```
Comment (utzcoz) [inline robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java:550]: What about using name `drawRect_withRectF`?

File: robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java [hunk 8]
```diff
@@ -529,4 +530,36 @@ public void restoreToCount() {
 
     assertThat(canvas.getSaveCount()).isEqualTo(save1);
   }
+
+  @Test
+  public void drawRect_toCanvas() {
```
Comment (utzcoz) [inline robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java:535]: What about using name `drawRect_withPureFloatPosition`?

Comment (Akshay2131) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java:535]: ok, i do it..

File: robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java (hunk 7 above)
Comment (Akshay2131) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java:550]: i do it..

**Comment by utzcoz** [comment]
This PR also has format problem: https://github.com/robolectric/robolectric/runs/5805197715?check_suite_focus=true. It's recommended to follow https://github.com/robolectric/robolectric/wiki/Robolectric's-code-style to format code locally before pushing. And could you squash your commits to one and push it again? Because they do the one thing.

**Comment by utzcoz** [comment]
@Akshay2131 your new commit message is not recommended. After merged, the `Combined push` say nothing about it does. What about changing commit message to `Support Canvas#drawRect with RectF`?

File: robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java [hunk 9]
```diff
@@ -529,4 +530,36 @@ public void restoreToCount() {
...
+  @Test
+  public void draw_Rect() throws Exception {
+    Canvas canvas = new Canvas();
+    Paint paint = new Paint();
+    paint.setColor(Color.WHITE);
+    RectF rect = new RectF(2f, 4f, 10f, 10f);
+    canvas.drawRect(2f, 4f, 10f, 10f, paint);
+    ShadowCanvas shadowCanvas = shadowOf(canvas);
+
+    assertThat(shadowCanvas.getDrawnRect(0).left).isEqualTo(2f);
```
Comment (utzcoz) [inline robolectric/src/test/java/org/robolectric/shadows/ShadowCanvasTest.java:558]: What about extract the result of `shadowCanvas.getDrawnRect` to a local parameter? For example,

```Java
Rect drawnRect = shadowCanvas.getDrawnRect(0);
assertThat(drawnRect.left).isEqualsTo(2f);
assertThat(drawnRect.top).isEquals(4f);
...
```

File: robolectric/src/test/java/org/robolectric/shadows/ShadowTelephonyManagerTest.java [hunk 10]
```diff
@@ -242,9 +242,9 @@ public void onCellInfo(List<CellInfo> list) {
   }
 
   @Test
-  public void shouldGiveNetworkCountryIso() {
+  public void shouldGiveNetworkCountryIsoInLowercase() {
```
Comment (utzcoz) [inline robolectric/src/test/java/org/robolectric/shadows/ShadowTelephonyManagerTest.java:245]: IMO, it is not related to the problem of this PR fixed.

File: robolectric/src/test/java/org/robolectric/shadows/ShadowTelephonyManagerTest.java [hunk 11]
```diff
@@ -506,7 +506,7 @@ public void shouldGetSimIso() {
 
   @Test
   @Config(minSdk = N, maxSdk = Q)
-  public void shouldGetSimIosWhenSetUsingSlotNumber() {
+  public void shouldGetSimIsoWhenSetUsingSlotNumber() {
```
Comment (utzcoz) [inline robolectric/src/test/java/org/robolectric/shadows/ShadowTelephonyManagerTest.java:509]: IMO, it is not related too.

File: robolectric/src/test/java/org/robolectric/shadows/ShadowWindowManagerGlobalTest.java [hunk 12]
```diff
@@ -22,6 +23,11 @@
 @RunWith(AndroidJUnit4.class)
 public class ShadowWindowManagerGlobalTest {
 
+  @Before
+  public void setup() {
+    System.setProperty("robolectric.areWindowsMarkedVisible", "true");
```
Comment (utzcoz) [inline robolectric/src/test/java/org/robolectric/shadows/ShadowWindowManagerGlobalTest.java:28]: I remember it is related to another commit in google branch or opening PR.

**Review by utzcoz — CHANGES_REQUESTED** [review]

**Comment by utzcoz** [comment]
@Akshay2131 Looks like there are some changes not related to this PR, and similar other commits on google branch or opening PR, could you remove them?

File: robolectric/src/test/java/org/robolectric/shadows/ShadowWindowManagerGlobalTest.java (hunk 12 above)
Comment (Akshay2131) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowWindowManagerGlobalTest.java:28]: Actually, before pushing it again i pulled the changes to my branch and then pushed..
That's why other changes appeared

Comment (Akshay2131) [reply robolectric/src/test/java/org/robolectric/shadows/ShadowWindowManagerGlobalTest.java:28]: Let me fix it.

**Comment by Akshay2131** [comment]
I was having issues to remove extra added files here.. So, i created a new PR that is:  ` Support Canvas#drawRect with RectF #7210 `
Could you please review that and close this one.

**Comment by utzcoz** [comment]
Thanks @Akshay2131 . I will close this PR.
